### PR TITLE
Modules & Persistence

### PR DESCRIPTION
**Pull request by T-Dnzt**
This PR adds persistence to Optimus Prime. It also comes with a new architecture for modules (exception handling, persistence). Modules are loaded with the `ModuleLoader`. 
## Modules Injection

Modules can be loaded in a pipeline either in Ruby or in the YML configuration. All the modules are **optional**.

``` ruby
modules = {
      persistence: {
        options: { dsn: 'sqlite:test.db' }
      },
      exceptional: {
        adapter: 'Sentry',
        options: { dsn: 'test' }
      }
    }
OptimusPrime::Pipeline.new({}, :a_cool_name,  modules)
```

Or in a YML file:

``` yaml
test_pipeline:

  modules:
    persistence:
      options: 
        dsn: 'testdb'
    exceptional:
      adapter: Sentry
      options:
        dsn: 'test'

  graph:
    a:
    ...
```
## PUB/SUB Events

Events can be broadcasted from any step inside a pipeline by using:

``` ruby
broadcast(:event_name, any, number, of, arguments)
```

The listener(s) needs to implement the correct method to receive those:

``` ruby
def event_name(any, number, of, arguments)
end
```

The persistence module is listening to the following events:
- `pipeline_started` (broadcasted from Pipeline)
- `pipeline_finished` (broadcasted from Pipeline)
- `pipeline_failed` (broadcasted from Pipeline)
- `load_job_started` (broadcasted from CloudStorageToBigquery)
- `load_job_finished` (broadcasted from CloudStorageToBigquery)
- `load_job_failed` (broadcasted from CloudStorageToBigquery)
## Persistence Module

The persistence module writes pipelines and load jobs states inside a DB. Each pipeline can write to a different DB or to the same one, it does not matter. The DB `dsn` is passed as an option to a pipeline and used to instantiate `Persistence::Base`. This class will create a connection to the DB and run the migrations if the tables are not existing. 

Once setup, the persistence listener is subscribed to the pipeline and each of its steps. It will then listen to the events specified above and save the states to the DB. 

Both for pipeline and load job, the status is either `started`, `finished` or `failed`. A pipeline is automatically set to `failed` when an exception is raised. Load jobs status is defined by BigQuery responses.

This module comes with two tables: `operations` and `load_jobs`. Operations are basically pipelines and load jobs are Google Storage to BigQuery specific. Load jobs won't be used if a pipeline does not use this step. However, pipelines will always be inserted in the `operations` table.

Here are the descriptions of those tables:

Operations:
- id: Database id
- pipeline_id: Pipeline name
- start_time: Time when the start event was broadcasted
- end_time: Time when the finished or failed event was broadcasted
- status: started, finished, failed
- error: exception + stacktrace if any

Load Jobs:
- id: Database id
- identifier: unique job id, the uri of the first file passed to the job
- job_id: BigQuery id
- operation_id: Pipeline id
- uris: String containing all the uris separated by a comma
- category: Destination table
- status: started, finished, failed
- start_time: Time when the start event was broadcasted
- end_time: Time when the finished or failed event was broadcasted
- error: exception + stacktrace if any

DB can also be migrated using `rake db:migrate`.  See `db.rake` for more commands.
New migrations can be created with: `rake 'generate:migration[create_entities]'`
### Walkthrough
1. Pipeline starting
2. Loading modules
3. Persistence is loaded, connecting to DB and running migrations
4. Persistence Listener is subscribed to the pipeline and its steps
5. Pipeline broadcast the `pipeline_started` event
6. Listener receives the event and create a record in the `operations` table.
7. If an exception is raised, it is captured, the `failed` event is broadcasted and the exception is raised again.
8. If no exception occurs, the pipeline broadcast the events `finished`.
## Exceptional Module

The exceptional module works in the same way than before except for the configuration writing. It should be now included under `modules:` in the YML and be renamed to exceptional. 
## Important Things to know
- `ModularLoader`: Loads all injected modules and make them accessible to the pipeline. Also make a list of subscribers for the pipeline and all its steps.
- `Listener`: The listener in the persistence module receives a set of events described below.
- PUB/SUB: More listeners can be added in modules by adding them to the `subscribers` list in the module loader.
- `pipeline.operate` should be used instead of `start.join`.

### CODE REVIEW

**Comment by jorinvo** [comment]
Nice job! :+1: 

Do we need to update the pipelines in any way?

**Comment by T-Dnzt** [comment]
Yeah we need to update the pipelines if we want to use the persistence module. Just add the config to the YML.

**Comment by Anumodana** [comment]
It looks good :+1:

**Comment by sjfbo** [comment]
Looks good to me!

File: spec/supports/config/test-config-exceptions.yml [hunk 1]
```diff
@@ -4,10 +4,11 @@ source2: &pipe_source <%= -> { 'spec/supports/csv/local_csv_source_pipe.csv' }.c
...
 
-  errors:
-    adapter: Sentry
-    options:
-      dsn: 'test'
+  modules:
+    exceptional:
+      adapter: Sentry
+      options:
+        dsn: 'test'
```
Comment (Anumodana) [inline spec/supports/config/test-config-exceptions.yml]: Why did you wrap the single quote around this dsn?

File: lib/optimus_prime/modules/persistence/listener.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        def initialize(base)
+          @base = base
+          @pipeline_name = nil
+          @operation_id = nil
+          @jobs = {}
+        end
+
+        def pipeline_started(pipeline)
+          @pipeline_name = pipeline.name
```
Comment (Anumodana) [inline lib/optimus_prime/modules/persistence/listener.rb]: Why did you set @pipeline_name in #initialize and #pipeline_started ?  I think no one uses it.

**Pull request merged**